JavaScript_Weekly/571/fuite.md

```diff
@@ -0,0 +1,95 @@
+toc_sticky: true
+---
+# fuite: 监测web应用中内存泄漏的工具
+在我们日常开发中，很少会去调试内存泄漏的问题，虽然是有工具可以实现调试的，但是它调试起来很复杂繁琐，通常也不会准确地告诉我们:我们的系统为什么会出现内存泄漏。
+
+在我的印象中，JavaScript是一个自带垃圾回收机制的语言，应该很少会出现内存泄漏的问题。但是当我第一次开始研究内存泄漏时，却发现它却成为了内存泄漏的主要来源。随着研究的深入，我发现在单页面应用程序(Single Page Apps，简称 spa)上是普遍存在内存泄漏问题的，只是很少有人对这些应用做测试。
+
+很少有web开发人员会注意到内存泄漏的问题，通常是因为他们发现自己的网站正在变慢，变卡甚至崩溃，或者偶然间在内存工具中发现页面占用了N兆(甚至N千兆)的内存时，才会考虑系统是否存在内存泄漏。不过当他们注意到这些问题的时候，系统出现内存泄露的点可能就不止一个了。
+
+回顾我之前写过的[关于内存泄露的文章](https://nolanlawson.com/2020/02/19/fixing-memory-leaks-in-web-applications/)就会发现，我通常都会建议开发中通过Chrome的内存监测工具来进行调试，但是由于工具的复杂程度，大部分人都很难通过它来查找除内存泄漏的原因。所以我希望有一个CLI工具来实现这些复杂的操作。
+
+因此我写了一个工具`fuite (法语“ leak”)`。在这个CLI工具中，我们可以用通过它来分析任意的一个网址是否存在内存泄漏的问题。
+```
+npx fuite https://example.com
+```
+在执行命令时，`fuite`会默认网站是呈现给用户的一个SPA页面，之后会抓取页面中的内部链接(例如``/about ``或``/contact``)。获取到链接后，会按照下面步骤来执行：
+1. 点击链接
+2. 点击浏览器的返回按钮
+3. 重复操作来监测内存是否增长
+当`fuite`监测到内存异常时，会告知我们哪些对象可能会导致内存泄漏。
+```
+Test         : Go to /foo and back
+Memory change: +10 MB
+Leak detected: Yes
+Leaking objects:
+| Object            | # added | Retained size increase |
+| ----------------- | ------- | ---------------------- |
+| HTMLIFrameElement | 1       | +10 MB                 |
+Leaking event listeners:
+| Event        | # added | Nodes  |
+| ------------ | ------- | ------ |
+| beforeunload | 2       | Window |
+Leaking DOM nodes:
+DOM size grew by 6 node(s) 
+```
+为了实现这个功能，我在`fuite`中使用了博客文章中阐述的基本策略。他会启动Chrome来运行n次(默认7次),并检查是否有对象发生内存泄漏了n次(7,14,21等）。
+
+`fuite`同时会分析全部的 ``Array``, ``Object``,`` Map``, ``Set``，事件监听器以及整个DOM。假如一个数组在执行了7次后同时也增长了7次，那么这个对象就有可能发生了内存泄漏。
+## 测试正式网站
+在反复进行页面跳转并返回的操作时，会发现在许多spa应用中发现内存泄漏的问题。下面是我对10个前端主流框架的首页进行的测试结果,并发现他们都存在内存泄漏的问题。
+
+| 站点    | 是否发生泄漏 | 内部连接数 | 平均增长 | 最大增长数 |
+| ------- | ------------- | -------------- | -------------- | ---------- |
+| Site 1  | yes           | 8              | 27.2 kB        | 43 kB      |
+| Site 2  | yes           | 10             | 50.4 kB        | 78.9 kB    |
+| Site 3  | yes           | 27             | 98.8 kB        | 135 kB     |
+| Site 4  | yes           | 8              | 180 kB         | 212 kB     |
+| Site 5  | yes           | 13             | 266 kB         | 1.07 MB    |
+| Site 6  | yes           | 8              | 638 kB         | 1.15 MB    |
+| Site 7  | yes           | 7              | 1.37 MB        | 2.25 MB    |
+| Site 8  | yes           | 15             | 3.49 MB        | 4.28 MB    |
+| Site 9  | yes           | 43             | 5.57 MB        | 7.37 MB    |
+| Site 10 | yes           | 16             | 14.9 MB        | 186 MB     |
+
+在这种情况下，`内部链接`是指所测试的内部链接数目，`平均增长`是指每个链接的平均内存增长(即每次跳转页面并返回) ，而`最大增长`是指内部链接内存泄漏的最大值。值得注意的时，这些数据并不是一次的测试结果而是`fuite`测试了7次之后的结果。
+
+当我们要确定这些结果的真实性时,你可以使用Chrome DevTools的`Memory`工具。下面图片是我选择了测试中表现最差的站点，点击其中的一个链接，然后按下后退键。每次操作都保存一个堆得快照，然后重复操作。
+
+![74a805399e214e6a054c93ad6e9d4ba.jpg](https://p9-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/34e6836feb8d47018b63611dd73ae2ff~tplv-k3u1fbpfcp-watermark.image?)
+
+我们可以发现每次当点击页面的时候，内存大概都会增加6M左右。
```

Comment:
On this particular site, memory grows by about 6 MB every time you click a link and go back.
```suggestion
我们可以发现每次点击连接并返回时，内存都会增加约6MB。
```